server/v2/stf/stf_router.go

```diff
@@ -147,3 +143,36 @@ func msgTypeURL(msg gogoproto.Message) string {
 
 	return gogoproto.MessageName(msg)
 }
+
+var _ router.Service = (*Router)(nil)
+
+// Router implements the STF router for msg and query handlers.
+type Router struct {
+	handlers map[string]appmodulev2.Handler
+}
+
+func (r Router) CanInvoke(_ context.Context, typeURL string) error {
+	_, exists := r.handlers[typeURL]
+	if !exists {
+		return fmt.Errorf("%w: %s", ErrNoHandler, typeURL)
+	}
+	return nil
+}
+
+func (r Router) InvokeTyped(ctx context.Context, req, resp protoiface.MessageV1) error {
+	handlerResp, err := r.InvokeUntyped(ctx, req)
+	if err != nil {
+		return err
+	}
+	gogoproto.Merge(resp, handlerResp)
```

Comment:
This will panic on custom typed gogo fields when the message is backed by a go struct, for example:
```golang

func Test_CoinMerge(t *testing.T) {
	src := sdk.NewInt64Coin("atom", 10)
	dst := &sdk.Coin{}
	require.Panics(t, func() {
		gogoproto.Merge(dst, &src)
	})
}
```
Previously I had (quite horrendous)
```golang
	reflect.Indirect(reflect.ValueOf(resp)).Set(reflect.Indirect(reflect.ValueOf(res)))
```
But I don't see any other alternatives on the table for a working implementation of `InvokeTypes` given the constrains of gogoproto and golang.  We could remove `InvokeTyped` and just use `Invoke`, forcing callers to cast the returned type themselves.